Add "gnustep" to nixpkgs

matthewbauer opened the pull request:
###### Motivation for this change

This builds off of @ashalkhakov's work in #5197. This should be compatible with latest changes to nixpkgs and mergeable. That PR can be closed once this one is merged.

Basically, it adds "gnustep" as an attribute within nixpkgs as well as some basic apps.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ashalkhakov committed:
gnustep: init

Adding new library: gnustep-startup, which packages the core
libraries necessary for GNUstep: gnustep-make, gnustep-base,
gnustep-gui, gnustep-backend.

ashalkhakov committed:
gnustep: packageg libs separately

Packing gnustep libs separately, with no use of gnustep-startup. Also,
fixed a bug in WindowMaker package (some imaging dependencies were not supplied).

ashalkhakov committed:
gnustep: remove gnustep-startup

Removing gnustep-startup (not needed anymore). Adding Gorm and
ProjectCenter applications (these mostly work, provided the environment
is set up manually).

ashalkhakov committed:
gnustep: cleanup

Cleaning up. Adding GNUstep package builder for abstracting out GNUstep
compilation specifics (with thanks to GitHub user lethalman).

The rules for using build_gnustep_package are as simple: any
GNUstep-based package that the package being compiled depends upon are
to be put in [deps] (this is used for setting up a buildEnv), while
other dependencies are put into [buildInputs] as usual.

ashalkhakov committed:
gnustep: use gnustep_builder

Built ProjectCenter and Gorm with the new gnustep_builder. Had to go
back and fix a few things.

ashalkhakov committed:
gnustep: add setup hook for make

Adding a setup-hook for gnustep-make.

ashalkhakov committed:
gnustep: add gworkspace and systempreferences

Packaging some basic GNUstep apps: GWorkspace and SystemPreferences.

Unfortunately, GWorkspace doesn't work well, because gdomap, gdnc, gpbs
are not started. Also, there is some issue with fonts not being found.

ashalkhakov committed:
gnustep: fix gworkspace

Minor fix: just to make GWorkspace buildable as it is.

ashalkhakov committed:
gnustep: add nixos deamons

Adding basic daemons: gdomap and gdnc. It seems that GWorkspace
does is unable to work properly without the daemons.

ashalkhakov committed:
gnustep: cleanup

Major clean-up. Everything builds fine.

ashalkhakov committed:
gnustep: fix gdnc, gdomap

Both gdnc and gdomap seem to work.

matthewbauer committed:
gnustep: rescope everything into one folder

This should make merge conflicts easier to
handle. "gnustep" prefix has been removed to
make thing simpler. So "gnustep_make" is now
"make" within the gnustep scope.

matthewbauer committed:
gnustep: provide default meta

Meta data like maintainers, license, and homepage is shared throughout the
"gnustep" project. Everything going through "gsmakeDerivation" now
shares overridable metadata.

matthewbauer committed:
gnustep: update to latest upstream versions

back, base, gorm, gui, gworkspace, and make were updated. xcode was
removed because it's not building correctly.

matthewbauer committed:
gnustep: add alias for old gnustep-make

joachifm commented on nixos/modules/services/networking/gdomap.nix:
```diff
@@ -0,0 +1,53 @@
+        ";
+      };
+
+      pidfile = mkOption {
```
Is there a good reason for making this configurable? It is desirable to use the `RuntimeDirectory` directive and place the pid file there, if possible, or simply at the `/run` top-level if it needs to be readable by others for some reason.

joachifm replied:
I also wonder whether it'd be possible to make the daemon remain in the foreground instead?

joachifm commented on nixos/modules/services/x11/gdnc.nix:
```diff
@@ -0,0 +1,34 @@
+    systemd.services.gdnc = {
+      path = [ pkgs.gnustep_base ];
+      description = "gdnc: GNUstep distributed notification center";
+      requires = [ "gdomap.service" ];
```
Possibly should also have an ordering directive; dependencies without ordering implies parallel activation.

joachifm commented on pkgs/development/libraries/gnustep/back/default.nix:
```diff
@@ -0,0 +1,22 @@
+  };
+  buildInputs = [ cairo base gui freetype pkgconfig x11 ];
+  meta = {
+    description = "GNUstep-back is a generic backend for GNUstep.";
```
Please remove the package name and trailing punctuation.

joachifm replied:
These packages should specify license, platforms, homepage, and ideally also maintainers.

joachifm commented on pkgs/development/libraries/gnustep/gui/fixup-all.patch:
```diff
@@ -0,0 +1,486 @@
+ 
+ # Things to do after uninstalling
+ # after-uninstall::
+Only in gnustep-gui-0.24.0.patched: GNUmakefile.postamble~
```
This and the following lines like it look like junk to me, surely it is unnecessary for this patch to apply?

joachifm commented on nixos/modules/services/networking/gdomap.nix:
```diff
@@ -0,0 +1,53 @@
+  #
+  options = {
+    services.gdomap = {
+      enable = mkOption {
```
Note that `mkEnableOption` is a convenient short-hand for these types of option definitions.

joachifm commented on pkgs/development/libraries/gnustep/base/fixup-paths.patch:
```diff
@@ -0,0 +1,384 @@
+Only in gnustep-base-1.24.7: base.make
```
Looks like junk to me

matthewbauer committed:
gnustep: remove unneeded lines in patches

There were alot of unnecessary "Only In" lines that were left in the
patch files. This should remove all of them.

matthewbauer committed:
gnustep: fix naming of gnustep stuff

This should fix the NixOS issues.

joachifm commented on nixos/modules/services/networking/gdomap.nix:
```diff
@@ -11,22 +11,10 @@ in
-        default = "/tmp/gdomap.pid";
-	description = "Location of the pid file for gdomap daemon";
-      };
+      enable = mkEnableOption "Whether to enable gdomap, the GNUstep distributed objects daemon";
```
Note that `mkEnableOption` expects the parameter to be the name of the service; I believe this will expand to a description like "Whether to enable Whether to enable gdomap".

matthewbauer committed:
gnustep: remove gdnc service

gdnc is a user process and can't be made into a NixOS module very
easily. It can still be put in the user's login script. According to the
GNUstep documentation it will be started as soon as it is needed.

matthewbauer committed:
gnustep: fix gdomap service

This gets rid of the rest of the pidfile stuff and makes gdomap just act
like a normal systemd process. Also reword "enable" option.

matthewbauer committed:
gnustep: fixup descriptions, remove punctuation

This is based on @joachifm's review. Reworded description in each of
GNUstep's attributes.

matthewbauer committed:
libobjc2: 1.7 -> 1.8.1

garbas commented:
@matthewbauer i've restarted travis jobs and they are still failing. could you take a look?

matthewbauer committed:
gnustep: attempt to fix osx building

Set with-layout=fhs. This should prevent OS X from having compat issues.

matthewbauer committed:
gnustep: move to pkgs/desktops/gnustep

matthewbauer commented:
@garbas @joachifm 
This is now building alright on Linux. It should also be able to build on OS X but I've been having trouble with that. I've disabled OS X for now. Can we get this merged fairly soon?

joachifm commented:
I'm unable to reproduce the travis error. Let's do it.
